core/trino-main/src/main/java/io/trino/operator/join/SortedPositionLinks.java

```diff
@@ -109,31 +110,47 @@ public Factory build()
             ArrayPositionLinks.FactoryBuilder arrayPositionLinksFactoryBuilder = ArrayPositionLinks.builder(size);
             int[][] sortedPositionLinks = new int[size][];
 
-            for (Int2ObjectMap.Entry<IntArrayList> entry : positionLinks.int2ObjectEntrySet()) {
+            Iterator<Int2ObjectMap.Entry<IntArrayList>> iterator = positionLinks.int2ObjectEntrySet().fastIterator();
+            while (iterator.hasNext()) {
+                Int2ObjectOpenHashMap.Entry<IntArrayList> entry = iterator.next();
                 int key = entry.getIntKey();
                 IntArrayList positions = entry.getValue();
-                positions.sort(comparator);
 
-                sortedPositionLinks[key] = new int[positions.size()];
-                for (int i = 0; i < positions.size(); i++) {
-                    sortedPositionLinks[key][i] = positions.getInt(i);
-                }
-
-                // ArrayPositionsLinks.Builder::link builds position links from
-                // tail to head, so we must add them in descending order to have
-                // smallest element as a head
-                for (int i = positions.size() - 2; i >= 0; i--) {
-                    arrayPositionLinksFactoryBuilder.link(positions.getInt(i), positions.getInt(i + 1));
-                }
-
-                // add link from starting position to position links chain
-                if (!positions.isEmpty()) {
-                    arrayPositionLinksFactoryBuilder.link(key, positions.getInt(0));
+                int[] sortedPositions = positions.toIntArray();
+                sortedPositionLinks[key] = sortedPositions;
+
+                if (sortedPositions.length > 0) {
+                    // Single element arrays need not be sorted and have no internal pairs to link
+                    if (sortedPositions.length > 1) {
+                        // Use the positions elements array for the merge sort temporary work buffer to avoid an extra redundant
+                        // copy. This works because we know that initially it has the same values as the array being sorted
+                        IntArrays.mergeSort(sortedPositions, 0, sortedPositions.length, comparator, positions.elements());
```

Comment:
why not `quickSort`?

Reply:
A few reasons:
1. (Most importantly) this was already the algorithm being used, I'm just exposing the opportunity to use an existing copy as the working buffer instead of creating a new copy of the array internally but not trying to fundamentally alter the sorting algorithm characteristics already in place.
2. The implementation of `IntArrayList` chooses `mergeSort` over `quickSort` when passed an external comparator because it's a stable sorting algorithm
3. There's a specific optimization for `mergeSort` over `int[]` that allows "merging" two sorted sub-regions regions via `System.arraycopy` instead of individual index load / store operations.

Reply:
> The implementation of IntArrayList chooses mergeSort over quickSort when passed an external comparator because it's a stable sorting algorithm

I don't think it matters here really. There is also `IntArrays#unstableSort(int[], int, int, it.unimi.dsi.fastutil.ints.IntComparator)` that uses `quickSort`, which is potentially faster.

Could you run `io.trino.benchmark.BenchmarkInequalityJoin` to show perf differences?